woocommerce/google-listings-and-ads
# Retry Action Scheduler jobs on timeout

### Changes proposed in this Pull Request:



Closes #808.

This PR adds a new method to the `ActionSchedulerJobMonitor` service that monitors the async jobs and reschedules them when they timeout.

This method uses PHP's `register_shutdown_function` to register a callback to run on PHP script failure due to a timeout error. There weren't any flags set in the error to specify it's a timeout so we need to search for the string `Maximum execution time` on the last thrown PHP error message (fetched by `error_get_last`) to check if it's a timeout error. This might make it a bit less reliable but I couldn't find any cases of PHP changing that error message.


A new filter is introduced for the failure timeframe and the previous failure rate threshold filter is renamed to be similar to our other filter names.

I've also reduced the job failure threshold to 3 (previously 5) and set the failure timeframe to 2 hours instead of 1. This means that if a job fails 3 times in the past two hours we won't retry it. This is to prevent retrying the timed-out jobs since they take longer to fail. 

The failure monitor is also refactored a bit so that it only watches for the exact job with the given arguments instead of generally watching the job's processing hook. Previously when a job failed multiple times we wouldn't process it anymore even when it was run with other arguments. Now, we also check the job's arguments and process them if the arguments have changed. I had some doubt here though because in some cases when a job is failing, it could fail for any given arguments (e.g. not configured correctly or API errors, etc.), but in other cases, a job's failure could be just for that specific set of args (e.g. non-existing or invalid product provided).


### Detailed test instructions:

1. Add `while (true) {}` in the beginning of the `UpdateAllProducts::process_items` method
2. Sync all products (async) via the connection test page
3. Go to `WooCommerce > Status > Scheduled Actions`
4. Wait for the job to time out and confirm that it gets rescheduled once it times out
5. Repeat steps 2 to 4 for three other times and confirm that after the last attempt it will not get rescheduled
6. Add `while (true) {}` in the beginning of the `UpdateProducts::process_items` method
7. Sync one product (async) via the connection test page
8. Repeat steps 3 to 5 this time for the Update Products job



### Changelog entry

> Tweak - Retry async jobs on timeout
